Wrappers/Python/cil/optimisation/utilities/StepSizeMethods.py

```diff
@@ -138,3 +138,88 @@ def get_step_size(self, algorithm):
             raise ValueError(
                 'Could not find a proper step_size in {} loops. Consider increasing alpha or max_iterations.'.format(self.max_iterations))
         return self.alpha
+
+
+class BarzilaiBorweinStepSizeRule(StepSizeRule):
+
+    """ 
+    Applies the Barzilai- Borwein rule to calculate the step size (step_size).
+
+    Let :math:`\Delta x=x_k-x_{k-1}` and :math:`\Delta g=g_k-g_{k-1}`. A Barzilai-Borwein (BB) iteration is :math:`x_{k+1}=x_k-\alpha _kg_k` where the step size :math:`\alpha _k` is either
+
+    - :math:`\alpha _k^{LONG}=\frac{\Delta x\cdot\Delta x}{\Delta x\cdot\Delta g}`, or
+
+    - :math:`\alpha _k^{SHORT}=\frac{\Delta x \cdot\Delta g}{\Delta g \cdot\Delta g}`.
+    
+    Parameters
+    ----------
+    initial: float,
+        The step-size for the first iteration
+    mode: One of 'long', 'short' or 'alternate', default is 'short'. 
+        This calculates the step-size based on the LONG, SHORT or alternating between the two, starting with short. 
+    stabilisation_param: float or None, default is None 
+        In order to add stability the step-size has an upper limit of :math:`\Delta/\|g_k\|` were :math:`\Delta` is either a given positive constant, or determined to be the minimium of :math`:\Delta x: from the first 3 iterations. To "turn off" the stabilisation use `np.inf`.
+        
+    
+
+    Reference
+    ---------
+    - Barzilai, Jonathan; Borwein, Jonathan M. (1988). "Two-Point Step Size Gradient Methods". IMA Journal of Numerical Analysis. 8: 141–148
+    https://en.wikipedia.org/wiki/Barzilai-Borwein_method
+    
+    - Burdakov, O., Dai, Y.H. and Huang, N., 2019. Stabilized barzilai-borwein method. arXiv preprint arXiv:1907.06409.
+
+    - https://en.wikipedia.org/wiki/Barzilai-Borwein_method
+    """
+
+    def __init__(self, initial, mode='short', stabilisation_param=None):
+        '''Initialises the step size rule 
+        '''
+ 
+        self.mode=mode
+        self.store_grad=None 
+        self.store_x=None
+        self.initial=initial
+        if stabilisation_param is None:
+            self.adaptive = True
+            stabilisation_param =numpy.inf
+        else:
+            self.adaptive = False 
+            if not ( isinstance(stabilisation_param, Number) and stabilisation_param >=0):
+                raise TypeError(" The stabilisation_param should be None, a positive number or np.inf")
+        self.stabilisation_param=stabilisation_param
+        
+    
+
+    def get_step_size(self, algorithm):
+        """
+        Applies the B-B rule to calculate the step size (`step_size`)
+
+        Returns
+        --------
+        the calculated step size:float
+
+        """
+        if self.store_x is None:
```

Comment:
Could you add some comments on each of the parts of this algorithm so to make clearer what each of the steps achieves?